src/ILLink.RoslynAnalyzer/ITypeSymbolExtensions.cs

```diff
@@ -61,9 +61,11 @@ public static bool IsTypeOf (this ITypeSymbol symbol, string @namespace, string
 		public static bool IsTypeOf (this ITypeSymbol symbol, WellKnownType wellKnownType)
 		{
 			if (wellKnownType.TryGetSpecialType (out var specialType)) {
+				// Nullables are SpecialType.None, but the originalDefinition is SpecialType.Nullable_T
+				var symbolSpecialType = symbol.SpecialType == SpecialType.None ? symbol.OriginalDefinition.SpecialType : symbol.SpecialType;
```

Comment:
Substituted types are never special types, so rather than try to fix this up, I would make this a part of the API, meaning that the API expects to always receive an original definition. You could assert this by saying

```
Debug.Assert(symbol.Equals(symbol.OriginalDefinition, SymbolEqualityComparer.Default));
```

Reply:
If we always want the SpecialType of the OriginalDefinition, would it make more sense to do something like the following? This way the caller doesn't need to remember to use the OriginalDefinition.
```C#
var symbolSpecialType = symbol.OriginalDefinition.SpecialType;
```

Reply:
That would always work in getting the well-known type, but I'm not sure if it would create a trap. That is, at the call-site, does the code then use the substituted type or the original definition? If the code requires the original definition to work properly, it's probably best to just require its use, and catch any logic errors early.